### PR TITLE
Modify Generator to allow passing a logits processor to it

### PR DESCRIPTION
**Pull request by RobinPicard**
Addresses #1487 and #1414 

This PR modifies the `Generator` classes. Instead of accepting only an `output_type` argument, which could be a python type or an Interefular FSM, it now accepts 3 distinct arguments: `output_type`, `fsm` and `logits_processor`. Additionally, the PR creates tests for the `Generator` objects.

### CODE REVIEW

File: outlines/generator.py [hunk 1]
```diff
@@ -91,8 +106,51 @@ def stream(self, prompt, **inference_kwargs):
 def Generator(
     model: Union[SteerableModel, BlackBoxModel],
     output_type: Optional[Any] = None,
+    *,
+    logits_processor: Optional[OutlinesLogitsProcessor] = None,
```
Comment (rlouf) [inline outlines/generator.py:110]: Can we call it `processors`? I think in this context it makes sense.

File: outlines/generator.py [hunk 2]
```diff
@@ -91,8 +106,51 @@ def stream(self, prompt, **inference_kwargs):
...
+    fsm: Optional[interegular.fsm.FSM] = None,
```
Comment (rlouf) [inline outlines/generator.py:111]: I would prefer a separate output type instead of a kwarg here.

Comment (RobinPicard) [reply outlines/generator.py:111]: So keep what we had before, right?

Comment (rlouf) [reply outlines/generator.py:111]: Yes we could have a `FSM` object like we have `CFG` or `JsonSchema`
